pkg/reconciler/delivery/modify.go

```diff
@@ -0,0 +1,114 @@
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
+	listers "knative.dev/serving/pkg/client/listers/serving/v1"
+)
+
+/**************************************************************************************************************** 
+   modifyRouteSpec assigns traffic to all active Revisions according to the algorithm on page 8 of go/mydesigndoc
+   arguments:
+   - route: the current Route object
+   - r: a lister to query the Revisions by name
+   - newRevName: name string of the latest ready Revision
+   - policy: pointer to the Policy struct that commands the rollout process
+   return values:
+   - 1st value: a new route object whose spec field has been written with the desired state
+   - 2nd value: error if anything goes wrong
+****************************************************************************************************************/ 
+func modifyRouteSpec(route *v1.Route, r listers.RevisionLister, newRevName string, policy *Policy) (*v1.Route, error) {
+	// assumption 1: the current Route Status traffic % are all non-zero (any zero entries would not have been written)
+	// assumption 2: the current Route Status traffic entries are ordered from oldest to newest Revision
+	// first identify whether or not newRevName is already in the pool
+	nameListed := false
+	for _, t := range route.Status.Traffic {
+		if t.RevisionName == newRevName {
+			nameListed = true
+			break
+		}
+	}
+
+	// make a slice container to hold the new traffic assignments, and an ordered, lightweight roster of the pool
+	// that contains all current Revision names, INCLUDING the newest one
+	ln := len(route.Status.Traffic)
+	if !nameListed {
+		ln = ln + 1
+	}
+	if ln == 1 {
+		route.Spec.Traffic = []v1.TrafficTarget{{
+			ConfigurationName: route.Name, // assuming namespace/name matches for Route & Config
+			LatestRevision: ptr.Bool(true),
+			Percent: ptr.Int64(100),
+		}}
+		return route, nil
+	}
+	traffic := make([]v1.TrafficTarget, ln) // container for holding traffic assignments
+	roster := make([]string, ln)            // ordered list of all Revision names in the pool
+	for i, t := range route.Status.Traffic {
+		roster[i] = t.RevisionName
+	}
+	if len(route.Status.Traffic) < len(roster) {
+		roster[len(roster) - 1] = newRevName
+	}
+	
+	// go through the roster in reverse order (newest to oldest) and assign traffic to each Revision
+	alreadyAssigned := 0
+	for i := len(roster) - 1; i >= 0; i-- {
+		revision, err := r.Revisions(route.Namespace).Get(roster[i])
+		if err != nil {
+			return route, err
+		}
+		// exception for the first ever Revision
+		if revision.Labels[RevisionGenerationKey] == "1" {
+			traffic[i] = v1.TrafficTarget{
+				RevisionName: roster[i],
+				LatestRevision: ptr.Bool(false),
+				Percent: ptr.Int64(int64(100 - alreadyAssigned)),
+			}
+			break
+		}
+		timeElapsed := time.Now().Sub(revision.CreationTimestamp.Time)
+		want := computeNewPercentExplicit(policy, timeElapsed)
+		actual := min(want, 100 - alreadyAssigned)
```

Comment:
Note, here and elsewhere you presume that the total traffic sum is 100, but I might split traffic:
80% - config
20% - specific revision.
Then you can only shift 80% of the traffic

Reply:
Hmmm, is this a valid use case? My assumption is that the user intends to rollout any revisions they create. Seems like in this example the 20% specific revision here will remain static, independent from the rest, which doesn't obey that assumption.

Our current design doesn't accommodate this. If we want to allow use cases like this, then we might need to rethink the whole design.

Reply:
Yes, this is a valid case. For A/B testing, or for a long running experiments. E.g. it's quite popular to redirect 1% of traffic forever to an experimental revision and 99% for actual production. So when I rollout new production only 99% need to rolled over.

Reply:
I see what you mean here. If we want to support this, then I think we need a mechanism for the user to specify which revision is the "forever" one. I will talk to Thomas about this and address this in a separate PR; it's probably too much to squeeze it into this already huge PR :)